src/data/pnat.lean

```diff
@@ -24,51 +28,669 @@ def succ_pnat (n : ℕ) : ℕ+ := ⟨succ n, succ_pos n⟩
+      rw[mul_one] at h',exact ⟨h',le_refl k.val⟩}},
+  {exact ⟨nat.mod_le m.val k.val,le_of_lt (nat.mod_lt m.val k.property)⟩}
+end
+
+instance : has_dvd ℕ+ := ⟨λ k m, (k : ℕ) ∣ (m : ℕ)⟩
+
+theorem dvd_iff {k m : ℕ+} : k ∣ m ↔ (k : ℕ) ∣ (m : ℕ) := by {refl}
+
+theorem dvd_iff' {k m : ℕ+} : k ∣ m ↔ mod m k = k :=
+begin
+  change k.val ∣ m.val ↔ mod m k = k,
+  rw[nat.dvd_iff_mod_eq_zero],split,
+  {intro h,apply eq,rw[mod_coe,if_pos h],refl},
+  { intro h,by_cases h' : m.val % k.val = 0,
+    { exact h'},
+    { replace h : ((mod m k) : ℕ) = (k : ℕ) := congr_arg _ h,
+      rw[mod_coe,if_neg h'] at h,
+      exact (ne_of_lt (nat.mod_lt m.val k.property) h).elim}}
+end
+
+def div_exact {m k : ℕ+} (h : k ∣ m) : ℕ+ :=
+ ⟨(div m k).succ,nat.succ_pos _⟩
+
+theorem mul_div_exact {m k : ℕ+} (h : k ∣ m) : k * (div_exact h) = m :=
+begin
+ apply eq,rw[mul_coe],
+ change (k : ℕ) * (div m k).succ = m,
+ rw[mod_add_div m k,dvd_iff'.mp h,nat.mul_succ,add_comm],
+end
+
+theorem dvd_iff'' {k n : ℕ+} : k ∣ n ↔ ∃ m, k * m = n :=
+⟨λ h, ⟨div_exact h,mul_div_exact h⟩,
+ λ ⟨m,h⟩, dvd.intro (m : ℕ)
+          ((mul_coe k m).symm.trans (congr_arg subtype.val h))⟩
+
+theorem dvd_intro {k n : ℕ+} (m : ℕ+) (h : k * m = n) : k ∣ n :=
+ dvd_iff''.mpr ⟨m,h⟩
+
+theorem dvd_refl (m : ℕ+) : m ∣ m := dvd_intro 1 (mul_one m)
+
+theorem dvd_antisymm {m n : ℕ+} : m ∣ n → n ∣ m → m = n :=
+λ hmn hnm, subtype.eq (nat.dvd_antisymm hmn hnm)
+
+theorem dvd_trans {k m n : ℕ+} : k ∣ m → m ∣ n → k ∣ n :=
+@_root_.dvd_trans ℕ _ k.val m.val n.val
+
+theorem one_dvd (n : ℕ+) : 1 ∣ n := dvd_intro n (one_mul n)
+
+theorem dvd_one_iff (n : ℕ+) : n ∣ 1 ↔ n = 1 :=
+ ⟨λ h,dvd_antisymm h (one_dvd n),λ h, h.symm ▸ (dvd_refl 1)⟩
+
+def gcd (n m : ℕ+) : ℕ+ :=
+ ⟨nat.gcd n.val m.val,nat.gcd_pos_of_pos_left m.val n.pos⟩
+
+def lcm (n m : ℕ+) : ℕ+ :=
+ ⟨nat.lcm n.val m.val,
+  by {let h := mul_pos n.property m.property,
+      rw[← gcd_mul_lcm n.val m.val,mul_comm] at h,
+      exact pos_of_dvd_of_pos (dvd.intro (nat.gcd n.val m.val) rfl) h}⟩
+
+@[simp] theorem gcd_val (n m : ℕ+) : (gcd n m).val = nat.gcd n.val m.val := rfl
+
+@[simp] theorem lcm_val (n m : ℕ+) : (lcm n m).val = nat.lcm n.val m.val := rfl
+
+theorem gcd_dvd_left (n m : ℕ+) : (gcd n m) ∣ n := nat.gcd_dvd_left n.val m.val
+
+theorem gcd_dvd_right (n m : ℕ+) : (gcd n m) ∣ m := nat.gcd_dvd_right n.val m.val
+
+theorem dvd_gcd {m n k : ℕ+} (hm : k ∣ m) (hn : k ∣ n) : k ∣ gcd m n :=
+ @nat.dvd_gcd m.val n.val k.val hm hn
+
+theorem dvd_lcm_left  (n m : ℕ+) : n ∣ lcm n m := nat.dvd_lcm_left  n.val m.val
+
+theorem dvd_lcm_right (n m : ℕ+) : m ∣ lcm n m := nat.dvd_lcm_right n.val m.val
+
+theorem lcm_dvd {m n k : ℕ+} (hm : m ∣ k) (hn : n ∣ k) : lcm m n ∣ k :=
+ @nat.lcm_dvd m.val n.val k.val hm hn
+
+theorem gcd_mul_lcm (n m : ℕ+) : (gcd n m) * (lcm n m) = n * m :=
+ subtype.eq (nat.gcd_mul_lcm n.val m.val)
+
+def prime (p : ℕ+) : Prop := p.val.prime
 
 end pnat
+
+/-- The type of prime numbers
+-/
+def nat.primes := {p : ℕ // p.prime}
```

Comment:
It would be good to move this to `data/nat/prime`.

Reply:
OK, thanks.  I'm not very clear how this process is supposed to work.  If I make this change on my own PC and push it to the pnat branch of my mathlib fork, does it automatically get applied to this PR?  Is there a good guide to this kind of thing anywhere?  (I can get people to show me next week in Edinburgh if necessary.)

Reply:
It should just work. Make the changes, commit to your branch, and push.

Reply:
OK, thanks again.  I just added spaces after all the commas and I see that that got magically applied to the PR.  I'll do some further reorganization later.

Reply:
I have moved this to `data/nat/prime.lean`